source/Applications/Advanced/TransformPointCloudViaArucoMarker/TransformPointCloudViaArucoMarker.cpp

```diff
@@ -0,0 +1,313 @@
+            const int xRoundedDown = clamp(static_cast<int>(std::floor(x)), 0, width - 1);
+            const int xRoundedUp = clamp(static_cast<int>(std::ceil(x)), 0, width - 1);
+            const int yRoundedDown = clamp(static_cast<int>(std::floor(y)), 0, height - 1);
+            const int yRoundedUp = clamp(static_cast<int>(std::ceil(y)), 0, height - 1);
+
+            // Getting 3D coordinates for the four points next to the interpolation point
+            const auto q11 = points(yRoundedDown, xRoundedDown);
+            const auto q12 = points(yRoundedUp, xRoundedDown);
+            const auto q21 = points(yRoundedDown, xRoundedUp);
+            const auto q22 = points(yRoundedUp, xRoundedUp);
+
+            // Linear interpolation in x direction
+            // f(x,y1) = (x2 - x)/(x2 - x1) * f(q11) + (x - x1)/(x2 - x1) * f(q21)
+            const cv::Point3f fxy1{ (xRoundedUp - x) / (xRoundedUp - xRoundedDown) * q11.x
+                                        + (x - xRoundedDown) / (xRoundedUp - xRoundedDown) * q21.x,
+                                    (xRoundedUp - x) / (xRoundedUp - xRoundedDown) * q11.y
+                                        + (x - xRoundedDown) / (xRoundedUp - xRoundedDown) * q21.y,
+                                    (xRoundedUp - x) / (xRoundedUp - xRoundedDown) * q11.z
+                                        + (x - xRoundedDown) / (xRoundedUp - xRoundedDown) * q21.z };
+            // f(x,y2) = (x2 - x)/(x2 - x1) * f(q12) + (x - x1)/(x2 - x1) * f(q22)
+            const cv::Point3f fxy2{ (xRoundedUp - x) / (xRoundedUp - xRoundedDown) * q12.x
+                                        + (x - xRoundedDown) / (xRoundedUp - xRoundedDown) * q22.x,
+                                    (xRoundedUp - x) / (xRoundedUp - xRoundedDown) * q12.y
+                                        + (x - xRoundedDown) / (xRoundedUp - xRoundedDown) * q22.y,
+                                    (xRoundedUp - x) / (xRoundedUp - xRoundedDown) * q12.z
+                                        + (x - xRoundedDown) / (xRoundedUp - xRoundedDown) * q22.z };
+
+            // Linear interpolation in y direction
+            // f(x,y) = (y2 - y)/(y2 - y1) * f(x,y1) + (y - y1)/(y2 - y1) * f(x,y2)
+            markerPoints3D.emplace_back(cv::Point3f{ (yRoundedUp - y) / (yRoundedUp - yRoundedDown) * fxy1.x
+                                                         + (y - yRoundedDown) / (yRoundedUp - yRoundedDown) * fxy2.x,
+                                                     (yRoundedUp - y) / (yRoundedUp - yRoundedDown) * fxy1.y
+                                                         + (y - yRoundedDown) / (yRoundedUp - yRoundedDown) * fxy2.y,
+                                                     (yRoundedUp - y) / (yRoundedUp - yRoundedDown) * fxy1.z
+                                                         + (y - yRoundedDown) / (yRoundedUp - yRoundedDown) * fxy2.z });
+        }
```

Comment:
Don't have the time to review this math. Hopefully it's from a legit source and it's tested :)

Reply:
Correct. The sample is tested from a couple of different poses and it makes sense every time.